### PR TITLE
Adds a proof harness for aws_cryptosdk_cmm_generate_enc_materials

### PR DESCRIPTION
**Pull request by danielsn**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

**Comment by danielsn** [comment]
https://github.com/aws/aws-encryption-sdk-c/issues/539 to track the precondition cleanup

File: .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/Makefile [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+#########
+ABSTRACTIONS += $(SRCDIR)/c-common-helper-stubs/aws_hash_table_no_slots_override.c
+ABSTRACTIONS += $(SRCDIR)/c-common-helper-stubs/aws_string_destroy_override.c
+ABSTRACTIONS += $(SRCDIR)/c-common-helper-stubs/error.c
+ABSTRACTIONS += $(SRCDIR)/helper-src/openssl/bn_override.c
+ABSTRACTIONS += $(SRCDIR)/helper-src/openssl/ec_override.c
+ABSTRACTIONS += $(SRCDIR)/helper-src/openssl/evp_override.c
+
+# Enable after https://github.com/diffblue/cbmc/issues/5344 is fixed
+#AWS_DEEP_CHECKS = 1
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/Makefile:38]: Do we have an issue here to update the proofs with this comment?

File: .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/aws_cryptosdk_cmm_generate_enc_materials_harness.c [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+ * On failure returns AWS_OP_ERR, sets address pointed to by output to NULL, and sets
+ * internal AWS error code.
+ */
+int generate_enc_materials(
+    struct aws_cryptosdk_cmm *cmm,
+    struct aws_cryptosdk_enc_materials **output,
+    struct aws_cryptosdk_enc_request *request) {
+    // assert(aws_cryptosdk_cmm_base_is_valid(cmm));
+    // assert(__CPROVER_w_ok(output, sizeof(*output)));
+    // assert(aws_cryptosdk_enc_request_is_valid(request));
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/aws_cryptosdk_cmm_generate_enc_materials_harness.c:53]: Why these preconditions are commented out? They should be added to the harness.

File: .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/aws_cryptosdk_cmm_generate_enc_materials_harness.c [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    ensure_cryptosdk_edk_list_has_allocated_list(&materials->encrypted_data_keys);
+    __CPROVER_assume(aws_cryptosdk_edk_list_is_valid(&materials->encrypted_data_keys));
+
+    // Stub until https://github.com/diffblue/cbmc/issues/5344 is fixed
+    // edk_list Precondition: The list has valid list elements
+    /*
+      __CPROVER_assume(aws_cryptosdk_edk_list_elements_are_bounded(&materials->encrypted_data_keys, MAX_STRING_LEN));
+      ensure_cryptosdk_edk_list_has_allocated_list_elements(&materials->encrypted_data_keys);
+      __CPROVER_assume(aws_cryptosdk_edk_list_elements_are_valid(&materials->encrypted_data_keys));
+    */
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/aws_cryptosdk_cmm_generate_enc_materials_harness.c:89]: We should have an issue here to update these proofs once #5344 is fixed.

File: .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/aws_cryptosdk_cmm_generate_enc_materials_harness.c [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    struct aws_cryptosdk_enc_materials **output = can_fail_malloc(sizeof(*output));
+    __CPROVER_assume(output);
+
+    // Run the function under test.
+    if (aws_cryptosdk_cmm_generate_enc_materials(cmm, output, request) == AWS_OP_SUCCESS) {
+        assert(aws_cryptosdk_enc_materials_is_valid(*output));
+    } else {
+        assert(*output == NULL);
+    }
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/aws_cryptosdk_cmm_generate_enc_materials_harness.c:134]: We should also check that `cmm` and `request` are valid.
```c
assert(aws_cryptosdk_cmm_base_is_valid(cmm));
assert(aws_cryptosdk_enc_request_is_valid(request));
```

File: .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/Makefile (hunk 1 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/Makefile:38]: https://github.com/aws/aws-encryption-sdk-c/issues/529

File: .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/aws_cryptosdk_cmm_generate_enc_materials_harness.c (hunk 3 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/aws_cryptosdk_cmm_generate_enc_materials_harness.c:89]: https://github.com/aws/aws-encryption-sdk-c/issues/529

File: .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/aws_cryptosdk_cmm_generate_enc_materials_harness.c (hunk 2 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/aws_cryptosdk_cmm_generate_enc_materials_harness.c]: fixed

File: .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/aws_cryptosdk_cmm_generate_enc_materials_harness.c (hunk 4 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_cmm_generate_enc_materials/aws_cryptosdk_cmm_generate_enc_materials_harness.c:134]: done

**Pull request merged**